### ISSUE TITLE
Client Certificate Support

### ISSUE THREAD

**Issue by WatcherWhale** [jthomperoo/custom-pod-autoscaler#164]
**Is your feature request related to a problem? Please describe.**
Add the ability to use client certificates, to authorize with in the http method for metric gathering.

**Describe the solution you'd like**
I want to specify a path where a `client.crt,` `client.key` and a `ca.crt` to use  in the HTTP request.

**Describe alternatives you've considered**
An alternative could be to use a Kubernetes TLS secret. 

**Additional context**
Here is an example of how to do it in GO https://smallstep.com/hello-mtls/doc/client/go.

**Comment by jthomperoo** [comment]
I've implemented this functionality in the merge request linked to this issue - I'm going to take a while to test it out manually alongside the automated tests I've provided. If you could test this out too just to see if it fits your use case that would be really useful.

This is available in the dev build with this tag: `0a67fa8`

You can use it like this:

```
FROM custompodautoscaler/python:0a67fa8
...
```

Thanks!

**Comment by WatcherWhale** [comment]
Is there a way to annotate the autoscaler pods themselves? That is how I inject the certificate into it.

**Comment by jthomperoo** [comment]
Sorry I'm not entirely sure I understand your question - do you mean can you provide this configuration directly with the pod definition for the autoscaler?

If so you can do that by providing the configuration in the CPA YAML:

```yaml
name: metric
value: |
  type: "http"
  timeout: 2500
  http:
    method: "GET"
    url: "https://www.custompodautoscaler.com"
    successCodes:
      - 200
    headers:
      exampleHeader: exampleHeaderValue
    parameterMode: query
    caCert: "/ca.pem"
    clientCert: "/client.cert"
    clientKey: "/client.key"
```

(Details on that in the docs here: https://custom-pod-autoscaler.readthedocs.io/en/latest/reference/configuration/)

If that's not what you mean let me know!

**Comment by WatcherWhale** [comment]
I use autocert from smallstep to provision the certificates into a pod and I need to annotate the pod itself like in a deployment:
```yml
apiVersion: apps/v1
kind: Deployment
metadata: {name: hello-mtls, labels: {app: hello-mtls}}
spec:
  replicas: 1
  selector: {matchLabels: {app: hello-mtls}}
  template:
    metadata:
      annotations:
        # AUTOCERT ANNOTATION HERE -v ###############################
        autocert.step.sm/name: hello-mtls.default.svc.cluster.local #
        # AUTOCERT ANNOTATION HERE -^ ###############################
      labels: {app: hello-mtls}
    spec:
      containers:
      - name: hello-mtls
        image: smallstep/hello-mtls-server-go:latest
```
 Source: https://github.com/smallstep/autocert#annotate-pods-to-get-certificates
 
 Can I also do this with a CustomPodAutoscaler? Maybe this issue is the wrong place to ask this :sweat_smile:

**Comment by jthomperoo** [comment]
Hey, sorry again for the slow response - I just get to these issues when I get a bit of free time. 

So for your certs through annotations that _should_ be supported by the Custom Pod Autoscaler as long as the certs end up somewhere on the pod as a file you can reference, unfortunately at the minute there's a bug with the operator for the Custom Pod Autoscaler that does not correctly copy across metadata, so to get this to work this bug will need fixed.

The bug is jthomperoo/custom-pod-autoscaler-operator#87 - hopefully I can get round to fixing that soon, the person who raised it has done a bit of digging and very nicely found a solution that should work.

**Comment by jthomperoo** [comment]
This is available in [`v2.6.0`](https://github.com/jthomperoo/custom-pod-autoscaler/releases/tag/v2.6.0).
